agent/acs/handler/payload_responder.go

```diff
@@ -0,0 +1,289 @@
+			loggerfield.MessageID: aws.StringValue(payload.MessageId),
+		})
+		return
+	}
+
+	// Only need to stop the task; it brings down the containers too.
+	taskEvent := api.TaskStateChange{
+		TaskARN: *task.Arn,
+		Status:  apitaskstatus.TaskStopped,
+		Reason:  UnrecognizedTaskError{err}.Error(),
+		// The real task cannot be extracted from payload message, so we send an empty task.
+		// This is necessary because the task handler will not send an event whose
+		// Task is nil.
+		Task: &apitask.Task{},
+	}
+
+	tmHandler.taskHandler.AddStateChangeEvent(taskEvent, tmHandler.ecsClient)
+}
+
+// addTasks adds the tasks to the task engine based on the skipAddTask condition.
+// This is used to add non-stopped tasks before adding stopped tasks.
+func (tmHandler *taskMessageHandler) addTasks(payload *ecsacs.PayloadMessage, tasks []*apitask.Task,
+	skipAddTask skipAddTaskComparatorFunc) ([]*ecsacs.IAMRoleCredentialsAckRequest, bool) {
+	allTasksOK := true
+	var credentialsAcks []*ecsacs.IAMRoleCredentialsAckRequest
+	for _, task := range tasks {
+		if skipAddTask(task.GetDesiredStatus()) {
+			continue
+		}
+		tmHandler.taskEngine.AddTask(task)
+		// Only need to save task to DB when its desired status is RUNNING (i.e. this is a new task that we are going
+		// to manage). When its desired status is STOPPED, the task is already in the DB and the desired status change
+		// will be saved by task manager.
+		if task.GetDesiredStatus() == apitaskstatus.TaskRunning {
+			err := tmHandler.dataClient.SaveTask(task)
+			if err != nil {
+				logger.Error("Failed to save data for task", logger.Fields{
+					loggerfield.TaskARN: task.Arn,
+					loggerfield.Error:   err,
+				})
+				allTasksOK = false
+			}
+		}
+
+		ackCredentials := func(id string, description string) {
+			ack, err := tmHandler.ackCredentials(payload.MessageId, id)
+			if err != nil {
+				allTasksOK = false
+				logger.Error(fmt.Sprintf("Failed to acknowledge %s credentials for task",
+					description), logger.Fields{
+					"task":            task.String(),
+					loggerfield.Error: err,
+				})
+				return
+			}
+			credentialsAcks = append(credentialsAcks, ack)
+		}
+
+		// Generate an ack request for the credentials in the task, if the
+		// task is associated with an IAM role or the execution role.
+		taskCredentialsID := task.GetCredentialsID()
+		if taskCredentialsID != "" {
+			ackCredentials(taskCredentialsID, "task iam role")
+		}
+
+		taskExecutionCredentialsID := task.GetExecutionCredentialsID()
+		if taskExecutionCredentialsID != "" {
+			ackCredentials(taskExecutionCredentialsID, "task execution role")
+		}
+	}
+	return credentialsAcks, allTasksOK
+}
+
+func (tmHandler *taskMessageHandler) ackCredentials(messageID *string, credentialsID string) (
+	*ecsacs.IAMRoleCredentialsAckRequest, error) {
+	creds, ok := tmHandler.credentialsManager.GetTaskCredentials(credentialsID)
+	if !ok {
+		return nil, fmt.Errorf("credentials could not be retrieved")
```

Comment:
Nit:Should these errors not be complete sentences?

Reply:
In Golang it is idiomatic for error strings not to be capitalized and not end with punctuation. I believe this is a convention routinely followed in our code base.

See here for reference: https://google.github.io/styleguide/go/decisions.html#error-strings